torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -4064,6 +4067,42 @@ Tensor linalg_det_backward(
   }
 }
 
+std::tuple<Tensor, Tensor, Tensor> attn_backward(
+    const Tensor& grad_o,
+    const Tensor& grad_a,
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v) {
+  if (!grad_o.defined() && !grad_a.defined()) {
+    return {};
+  }
+
+  Tensor a = at::tanh(at::matmul(q, k.transpose(0, 1)));
+
+  if (grad_o.defined()) {
+    Tensor q_o =
+        at::matmul(at::matmul(grad_o, v.transpose(0, 1)) * (1 - a.pow(2)), k);
+    Tensor k_o = at::matmul(
+        (at::matmul(grad_o, v.transpose(0, 1)) * (1 - a.pow(2)))
+            .transpose(0, 1),
+        q);
+    Tensor v_o = at::matmul(a.transpose(0, 1), grad_o);
+
+    if (!grad_a.defined()) {
+      return std::tuple<Tensor, Tensor, Tensor>(q_o, k_o, v_o);
+    }
+    Tensor q_a = at::matmul(grad_a * (1 - a.pow(2)), k);
+    Tensor k_a = at::matmul((grad_a * (1 - a.pow(2))).transpose(0, 1), q);
+    return std::tuple<Tensor, Tensor, Tensor>(q_o + q_a, k_o + k_a, v_o);
+  }
+
+  // only grad_a is defined
+  Tensor q_a = at::matmul(grad_a * (1 - a.pow(2)), k);
+  Tensor k_a = at::matmul((grad_a * (1 - a.pow(2))).transpose(0, 1), q);
+  return std::tuple<Tensor, Tensor, Tensor>(
+      q_a, k_a, at::_efficientzerotensor(v.sizes(), v.options()));
```

Comment:
Cool use of zero-tensor. Returning an undefined tensor can also work here: `Tensor()`.